gammapy/maps/geom.py

```diff
@@ -631,29 +631,79 @@ def energy_mask(self, energy_min=None, energy_max=None, round_to_edge=False):
-        energy_axis = self.axes["energy"]
+    def create_mask(self, axis_name, edge_min=None, edge_max=None, round_to_edge=False):
+        """Create a mask over a given axis.
 
-        if round_to_edge:
-            energy_min, energy_max = energy_axis.round([energy_min, energy_max])
+        The energy bin must be fully contained to be included in the mask.
 
-        # TODO: make this more general
-        shape = (-1, 1) if self.is_hpx else (-1, 1, 1)
-        energy_edges = energy_axis.edges.reshape(shape)
+        Parameters
+        ----------
+        edge_min, edge_max : `~astropy.units.Quantity` or float
+            Bounds of the mask. If the axis used for the mask has a unit, passing `~astropy.units.Quantity` is mandatory.
+        round_to_edge: bool, optional
+            Wether to round `energy_min` and `energy_max` to the closest axis bin value.
+            See `~gammapy.maps.MapAxis.round`. Default is False.
 
-        # set default values
-        energy_min = energy_min if energy_min is not None else energy_edges[0]
-        energy_max = energy_max if energy_max is not None else energy_edges[-1]
+        Returns
+        -------
+        mask : `~gammapy.maps.Map`
+            Map containing the mask. The geometry of the map is the same as the initial geometry.
+        """
+        from . import Map
+
+        axes_names = self.axes_names
+
+        try:
+            axis = self.axes[axis_name]
+        except KeyError:
+            raise ValueError(
+                f"Axis with name `{axis_name}` not in list of axis name: {axes_names}."
+            )
+        shape = tuple([-1] + [1 for n in axes_names if n != axis_name])
+        axis_edges = axis.edges.reshape(shape)
+        edge_min = edge_min if edge_min is not None else axis_edges[0]
+        edge_max = edge_max if edge_max is not None else axis_edges[-1]
+
+        if axis.unit != "":
+            if not (
+                isinstance(edge_min, u.Quantity) and isinstance(edge_max, u.Quantity)
+            ):
+                raise TypeError(
+                    f"`edge_min` and `edge_max` must be instance of `~astropy.units.Quantity` for  axis with name `{axis_name}` as unit of `{axis.unit}`."
+                )
+            if not (
+                edge_min.unit.is_equivalent(axis.unit)
+                and edge_max.unit.is_equivalent(axis.unit)
+            ):
+                raise u.UnitsError(
+                    f"`edge_min` and `edge_max` unit must be equivalent to unit `{axis.unit}` of axis with name `{axis_name}`."
+                )
+
+        if round_to_edge:
+            edge_min, edge_max = axis.round([edge_min, edge_max])
 
-        mask = (energy_edges[:-1] >= energy_min) & (energy_edges[1:] <= energy_max)
+        axes_names.reverse()
+        mask = (axis_edges[:-1] >= edge_min) & (axis_edges[1:] <= edge_max)
+        mask = np.reshape(
+            mask,
+            tuple(len(axis_edges[1:]) if n == axis_name else 1 for n in axes_names),
```

Comment:
Again a little helper function in `MapAxes` would be very useful here. Or just use `np.expand_dims()` which should be simpler.

Reply:
I don't think a method on `MapAxes` would help here as if you are using e.g.`WcsGeom`, `.axes` only gather the non spatial axes . But here we need every dimension to broadcast correctly the mask. 
I now use `expand_dims` however.